lib/pry/config/behavior.rb

```diff
@@ -76,7 +76,8 @@ def default
       #
       def [](key)
         key = key.to_s
-        key?(key) ? @lookup[key] : (@default && @default[key])
+        obj = key?(key) ? @lookup[key] : (@default && @default[key])
+        Pry::Config::Lazy === obj ? obj.call : obj
```

Comment:
the documentation for `Pry::Config::Lazy` (defined in config.rb) could be improved to demonstrate how Pry.lazy can be used with the config:
```ruby
[1] pry(main)> _pry_.config.prompt_name = Pry.lazy { rand(50) }
=> #<Pry::Config::Lazy:0x00007fdeffaa2c48 @block=#<Proc:0x00007fdeffaa2bf8@(pry):1>>
[2] 2(main)> 'foo'
=> "foo"
[3] 27(main)>
```
it only says it is a 'proc with a name' right now, which doesn't describe the use case.

Reply:
Why change `obj.is_a?(Pry::Config::Lazy)`?

Reply:
in Pry i generally prefer to use `===` over `is_a?` because BasicObject does not implement `is_a?`, and because `===` still works when `is_a?` has been redefined/removed from the object being queried.

Reply:
Those are valid points but are they relevant here? It is highly unlikely that Lazy will inherit BasicObject or redefine is_a?. is_a? is a lot more understandable so I’d rather keep it.

Reply:
updated

Reply:
@kyrylo 
i think i should revert this, i wasn't thinking clearly earlier, the change you suggested wouldn't allow BasicObject instances to be read as values:

```ruby
_pry_.config.foo = BasicObject.new
_pry_.config['foo'] # raises an error because we call `is_a?` on BasicObject.
```

i added a failing spec to illustrate:
https://github.com/pry/pry/pull/1890/commits/33e0434d1f80c509ff52b7635a7d2632508d9664

Reply:
* i reverted your suggestion: https://github.com/pry/pry/pull/1890/commits/b1ece0b9edd385134b319eb5f671fdb44840a0e4
* fixed the spec to use `Module#===` instead of `.class` 
(BasicObject doesn't implement `.class`): https://github.com/pry/pry/pull/1890/commits/8a41c864271c67b587167cec32c9c7db2f54690f

should be green now.

Reply:
Why would any option accept BasicObject as a value? If you assume that BasicObject is common, then half of the gems will stop working. I feel like we're overcomplicating this. Please reconsider.

Reply:
i don't see why we should not support BasicObject as value, especially if it is only because `is_a?` looks better. it's also true that `Pry::Config < Pry::BasicObject`, so BasicObject-like objects are already stored there i guess, eg:
```ruby
Pry.configure do |config|
  config.my_plugin = Pry::Config.from_hash({foo: 1, bar: 2})
end
```

but our Pry::BasicObject is less limited. 
anyway, i don't think aesthetics should win over functionality, a BasicObject should be able to be stored as a configuration value.

Reply:
i follow the mruby project from time to time, i saw this commit come up today:
https://github.com/mruby/mruby/commit/ff08856fe314faa4d16b4502c0960a3475387846#diff-78405d471eefe34180e1f45b1a8f47f5

`String#===` is being used in every case where `is_a?` could have been used instead.
it makes me wonder if using `===` is really that unusual and if the bias towards `is_a?` is unfounded. 

anyway, back to the discussion we are having....

>  I really struggle to see the value of this functionality. 

i have the same problem, in reverse, i do not see why we would intentionally not support BasicObject just because of a style preference for `is_a?`. it would be a regression because storing BasicObject as a value in the config has never been a problem before. 

> but can you provide any real-world examples where this is actually needed? 

i don't have any real world examples that use BasicObject as a configuration value but i do not see this as a reason for not supporting it. we are putting personal style before functionality here, that is silly.

Reply:
i also think when dealing with arbitrary objects that could be anything and querying their class/module, `===` is a safer choice, it avoids issues monkey patches might cause and it also works with BasicObject.

most important thing to mention here is that storing BasicObject in the config already works and has since forever, we are going to introduce a breaking change because `===` isn't liked as much as `is_a?` aesthetically, it seems like a misguided decision.

Reply:
The linked example modifies a core class. Of course a generic method on a primitive class is expected to support as much as possible since nobody knows where it can be used. This is not the case with Pry config values where we control what we want.

GitHub doesn't allow searching for `===` but `is_a?` shows more than 6M results.
https://github.com/search?l=Ruby&q=is_a%3F&type=Code

You want to support some functionality that has no use and you are feeling strongly about it. I don't see what exactly you defend. Our APIs won't be worse if we go with `is_a?`. I believe that supporting BasicObject stuff implies certain restrictions and opens a potential can of worms and headache for maintainers (that I don't even want to think of). It makes already complex Config even more complex and since you say we already opened it, then let it be your way.

I am most certain that if this is a breaking change, it will break no code because nobody would ever need passing BasicObjects. You want a lot of flexibility but I fail to see why.

Either way, let's wrap this up and move on.

Reply:
I don't want to "support some functionality that has no use".
i just don't want to break functionality for the sake of aesthetics or a personal style.

the "complexity" we are discussing here comes down to this:

```ruby
obj.is_a?(Pry::Config::Lazy)
```
or

```ruby
Pry::Config::Lazy === obj
```

to me both are as complex as each other, the only difference is one allows us to store any object in the config and the other does not. why would we choose to limit ourselves, for the sake of style or aesthetics?

> Our APIs won't be worse if we go with is_a?.

I disagree, compared to now they would be more fragile and unable to support storing BasicObject (which was always supported, it's not a new feature or one that takes much effort to support). 

> I believe that supporting BasicObject stuff implies certain restrictions and opens a potential can of worms and headache for maintainers

We always supported it, there has never been a maintenance burden doing that until now and that burden is more arguing why we should not break features or introduce fragility because the alternative looks better or is perceived as less confusing.

> The linked example modifies a core class. Of course a generic method on a primitive class is expected to support as much as possible since nobody knows where it can be used. This is not the case with Pry config values where we control what we want.

I disagree again, the config is OpenStruct-like first and foremost, it should be able to store _any_ value and we shouldn't enforce limitations because of personal style choices, which are never as important as functionality.

> You want a lot of flexibility but I fail to see why.

why not? why choose to write code that is fragile? just because it looks better in your text editor?

Reply:
btw OpenStruct has no issues storing BasicObject:
```
[2] pry(main)> o = OpenStruct.new foo: BasicObject.new
=> #<OpenStruct:0x3febaa8a7ca0>
[3] pry(main)> o.foo
=> #<BasicObject:0x3febaa8a7cf0>
[4] pry(main)>
```

i don't think we have a reason not to either.